File: arm_wasd/kinematics.py
import math

class Kinematics:
    def __init__(self, L1, L2):
        self._L1 = L1
        self._L2 = L2
        self.Position_ = {
            "RightlyJoint0": 0.0,
            "RightlyJoint1": 0.0,
            "RightlyJoint2": 0.0,
            "RightlyJoint3": 0.0,
            "LeftyJoint0": 0.0,
            "LeftyJoint1": 0.0,
            "LeftyJoint2": 0.0,
            "LeftyJoint3": 0.0,
            "Lefty": False,
            "Rightly": False
        }
        self.ForwardKinematics_ = {
            "X": 0.0,
            "Y": 0.0,
            "Z": 0.0
        }
        self.transformed_ = {
            "X": 0.0,
            "Y": 0.0,
            "Z": 0.0
        }
        self._Lsum = self._L1 + self._L2
        self._Ldiff = abs(self._L1 - self._L2)
        self._Alpha = 0.0
        self._Beta = 0.0
        self._Gamma = 0.0
        self._xy = 0.0

    def InverseKinematics(self, targetX, targetY, targetZ):
        x_y = math.sqrt(targetX * targetX + targetY * targetY)
        xy_z = math.sqrt(x_y * x_y + targetZ * targetZ)
        print(xy_z)

        self.Position_["Lefty"] = False
        self.Position_["Rightly"] = False

        if self._Ldiff <= xy_z <= self._Lsum:
            self._Alpha = math.acos((xy_z * xy_z + self._L1 * self._L1 - self._L2 * self._L2) / (2 * self._L1 * xy_z)) * 180.0 / math.pi
            print(self._Alpha)
            self._Beta = math.acos((self._L1 * self._L1 + self._L2 * self._L2 - xy_z * xy_z) / (2 * self._L1 * self._L2)) * 180.0 / math.pi
            print(self._Beta)
            self._Gamma = math.atan2(targetZ, x_y) * 180.0 / math.pi
            print(self._Gamma)
            print(math.atan(targetZ/x_y)* 180.0 / math.pi)
            self.Position_["LeftyJoint0"] = self.Position_["RightlyJoint0"] = math.atan2(targetY, targetX) * 180.0 / math.pi
            self.Position_["RightlyJoint1"] = self._Gamma - self._Alpha
            self.Position_["RightlyJoint2"] = 180.0 - self._Beta
            self.Position_["LeftyJoint1"] = self._Gamma + self._Alpha
            self.Position_["LeftyJoint2"] = self._Beta - 180.0

            if (0 <= self.Position_["LeftyJoint0"] <= 350 and 0 <= self.Position_["LeftyJoint1"] <= 170 and 0 <= self.Position_["LeftyJoint2"] <= 180):
                self.Position_["Lefty"] = True
            if (0 <= self.Position_["RightlyJoint0"] <= 350 and 0 <= self.Position_["RightlyJoint1"] <= 170 and 0 <= self.Position_["RightlyJoint2"] <= 180):
                self.Position_["Rightly"] = True

    def ForwardKinematics(self, Joint0, Joint1, Joint2):
        print(self._L1 * math.sin(Joint1 / 180.0 * math.pi))
        self.ForwardKinematics_["Z"] = self._L1 * math.sin(Joint1 / 180.0 * math.pi) + self._L2 * math.sin(Joint1 / 180.0 * math.pi + Joint2 / 180.0 * math.pi)
        self._xy = self._L1 * math.cos(Joint1 / 180.0 * math.pi) + self._L2 * math.cos(Joint1 / 180.0 * math.pi + Joint2 / 180.0 * math.pi)
        self.ForwardKinematics_["X"] = self._xy * math.cos(Joint0 / 180.0 * math.pi)
        self.ForwardKinematics_["Y"] = self._xy * math.sin(Joint0 / 180.0 * math.pi)

    def CoordinateTrans(self, originX, originY, originZ, X, Y, Z):
        self.transformed_["X"] = X - originX
        self.transformed_["Y"] = Y - originY
        self.transformed_["Z"] = Z - originZ

# Example usage:
kinematics = Kinematics(274,130)
kinematics.InverseKinematics(101, 0, 285)
print(kinematics.Position_)
kinematics.ForwardKinematics(0, 45, 90)
print(kinematics.ForwardKinematics_)
# kinematics.CoordinateTrans(originX_value, originY_value, originZ_value, X_value, Y_value, Z_value)

File: kinematics.cpp
#include <Arduino.h>

class kinematics{
  public:
    kinematics(float L1, float L2){
      _L1 = L1;
      _L2 = L2;
    };

    struct Position {
      //float Joint0, Joint1, Joint2, Joint3;
      float RightlyJoint0, RightlyJoint1, RightlyJoint2, RightlyJoint3;
      float LeftyJoint0, LeftyJoint1, LeftyJoint2, LeftyJoint3;
      bool Lefty, Rightly;
    } Position_;

    struct ForwardKinematicsOutput {
      float X, Y, Z;
    } ForwardKinematics_, transformed_;

    void InverseKinematics(float targetX, float targetY, float targetZ) {

      float x_y = sqrt(targetX * targetX + targetY * targetY);
      float xy_z = sqrt(x_y * x_y + targetZ * targetZ);

      Position_.Lefty = false;
      Position_.Rightly = false;

      if (xy_z <= _Lsum && xy_z >= _Ldiff) {
        _Alpha = acos((xy_z * xy_z + _L1 * _L1 - _L2 * _L2) / (2 * _L1 * xy_z)) * 180.0f / PI;
        _Beta = acos((_L1 * _L1 + _L2 * _L2 - xy_z * xy_z) / (2 * _L1 * _L2)) * 180.0f / PI;
        _Gamma = atan2(targetZ, x_y) * 180.0f / PI;
        Position_.LeftyJoint0 = Position_.RightlyJoint0 = atan2(targetY, targetX) * 180.0f / PI;
        Position_.RightlyJoint1 = _Gamma - _Alpha;
        Position_.RightlyJoint2 = 180.0f - _Beta;
        Position_.LeftyJoint1 = _Gamma + _Alpha;
        Position_.LeftyJoint2 = _Beta - 180.0f;

        if (-180 <= Position_.LeftyJoint0 <= 180 && 0 <= Position_.LeftyJoint1 <= 180 && 0 <= Position_.LeftyJoint2 <= 180){
          Position_.Lefty = true;
        } 
        if (-180 <= Position_.RightlyJoint0 <= 180 && 0 <= Position_.RightlyJoint1 <= 180 && 0 <= Position_.RightlyJoint2 <= 180){
          Position_.Rightly = true;
        }
      }
    };

    void ForwardKinematics(float Joint0, float Joint1, float Joint2) {
      ForwardKinematics_.Z = _L1 * sin(Joint1) + _L2 * sin(Joint1 + Joint2);
      _xy = _L1 * cos(Joint1) + _L2 * cos(Joint1 + Joint2);
      ForwardKinematics_.X = _xy * cos(Joint0);
      ForwardKinematics_.Y = _xy * sin(Joint0);
    };

    void CoordinateTrans(float originX, float originY, float originZ, float X, float Y, float Z) {
      transformed_.X = X - originX;
      transformed_.Y = Y - originY;
      transformed_.Z = Z - originZ;
    };

  private:
    float _L1, _L2;
    float _Lsum = _L1 + _L2;
    float _Ldiff = abs(_L1 - _L2);
    float _Alpha, _Beta, _Gamma;
    float _xy;
    //float _Theta;
};